File: backend/app/schemas/imports.py
"""Import, CSV profile and rule schemas. Amounts are integer cents."""

import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.transaction import BalanceOut

DateFormat = Literal[
    "YYYY-MM-DD",
    "MM/DD/YYYY",
    "DD/MM/YYYY",
    "MM/DD/YY",
    "DD/MM/YY",
    "DD.MM.YYYY",
    "YYYYMMDD",
    "MM-DD-YYYY",
]


class ProfileFields(BaseModel):
    delimiter: str = Field(default=",", min_length=1, max_length=1)
    has_header: bool = True
    skip_rows: int = Field(default=0, ge=0, le=100)
    date_column: int = Field(default=0, ge=0)
    date_format: DateFormat = "MM/DD/YYYY"
    amount_mode: Literal["single", "debit_credit"] = "single"
    amount_column: int | None = Field(default=None, ge=0)
    debit_column: int | None = Field(default=None, ge=0)
    credit_column: int | None = Field(default=None, ge=0)
    invert_sign: bool = False
    description_column: int = Field(default=1, ge=0)
    memo_column: int | None = Field(default=None, ge=0)


class ProfileIn(ProfileFields):
    name: str = Field(min_length=1, max_length=120)
    account_id: int | None = None


class ProfileOut(ProfileIn):
    model_config = ConfigDict(from_attributes=True)

    id: int


class ProfileListOut(BaseModel):
    items: list[ProfileOut]


#: The file travels as text inside JSON (D-078); 5 MB is far beyond any statement.
MAX_CONTENT = 5 * 1024 * 1024


class PreviewIn(BaseModel):
    content: str = Field(max_length=MAX_CONTENT)
    profile: ProfileFields


class PreviewRowOut(BaseModel):
    date: datetime.date
    amount_cents: int
    description: str
    memo: str


class ProblemOut(BaseModel):
    line: int
    reason: str


class PreviewOut(BaseModel):
    columns: list[str]
    rows: list[PreviewRowOut]
    problems: list[ProblemOut]


class StageIn(BaseModel):
    account_id: int
    filename: str = Field(min_length=1, max_length=255)
    content: str = Field(min_length=1, max_length=MAX_CONTENT)
    profile_id: int | None = None


class MatchOut(BaseModel):
    transaction_id: int
    date: datetime.date
    amount_cents: int
    payee_id: int | None
    memo: str | None


class BillSuggestionOut(BaseModel):
    occurrence_id: int
    name: str
    due_date: datetime.date


class StagedRowOut(BaseModel):
    id: int
    row_index: int
    date: datetime.date
    amount_cents: int
    raw_description: str
    raw_memo: str
    payee_id: int | None
    new_payee_name: str | None
    category_id: int | None
    memo: str | None
    disposition: Literal["import", "skip", "match"]
    is_duplicate: bool
    applied_rule_id: int | None
    match: MatchOut | None
    bill: BillSuggestionOut | None
    link_bill: bool
    created_transaction_id: int | None


class RowPatch(BaseModel):
    payee_id: int | None = None
    new_payee_name: str | None = Field(default=None, max_length=120)
    category_id: int | None = None
    memo: str | None = Field(default=None, max_length=500)
    disposition: Literal["import", "skip", "match"] | None = None
    link_bill: bool | None = None


class BatchOut(BaseModel):
    id: int
    account_id: int
    filename: str
    format: Literal["csv", "ofx", "qfx"]
    profile_id: int | None
    status: Literal["staged", "committed", "undone"]
    row_count: int
    imported_count: int
    duplicate_count: int
    matched_count: int
    parse_errors: list[str]
    created_at: datetime.datetime
    committed_at: datetime.datetime | None
    undone_at: datetime.datetime | None


class BatchDetailOut(BatchOut):
    rows: list[StagedRowOut]


class BatchListOut(BaseModel):
    items: list[BatchOut]


class BatchResultOut(BaseModel):
    batch: BatchOut
    balances: list[BalanceOut]


# -------------------------------------------------------------------------------- rules


class RuleIn(BaseModel):
    name: str = Field(default="", max_length=120)
    is_active: bool = True
    match_field: Literal["description", "memo"] = "description"
    match_type: Literal["contains", "starts_with", "equals", "regex"] = "contains"
    match_value: str = Field(min_length=1, max_length=500)
    amount_min_cents: int | None = None
    amount_max_cents: int | None = None
    account_id: int | None = None
    set_payee_id: int | None = None
    set_category_id: int | None = None
    set_memo: str | None = Field(default=None, max_length=500)


class RulePatch(BaseModel):
    name: str | None = Field(default=None, max_length=120)
    is_active: bool | None = None
    match_field: Literal["description", "memo"] | None = None
    match_type: Literal["contains", "starts_with", "equals", "regex"] | None = None
    match_value: str | None = Field(default=None, min_length=1, max_length=500)
    amount_min_cents: int | None = None
    amount_max_cents: int | None = None
    account_id: int | None = None
    set_payee_id: int | None = None
    set_category_id: int | None = None
    set_memo: str | None = Field(default=None, max_length=500)


class RuleOut(RuleIn):
    model_config = ConfigDict(from_attributes=True)

    id: int
    priority: int


class RuleListOut(BaseModel):
    items: list[RuleOut]


class MoveIn(BaseModel):
    offset: int


class RuleTestRowOut(BaseModel):
    date: datetime.date
    amount_cents: int
    raw_description: str
    raw_memo: str
    account_id: int


class RuleTestOut(BaseModel):
    checked: int
    matches: list[RuleTestRowOut]

File: frontend/src/api/imports.ts
import { apiFetch } from './client'
import type { Balance } from './transactions'

/** Bank imports, CSV profiles and rules (SPEC §11). Amounts are integer cents. */

export const DATE_FORMATS = [
  'MM/DD/YYYY',
  'DD/MM/YYYY',
  'YYYY-MM-DD',
  'MM/DD/YY',
  'DD/MM/YY',
  'DD.MM.YYYY',
  'YYYYMMDD',
  'MM-DD-YYYY',
] as const
export type DateFormat = (typeof DATE_FORMATS)[number]

export type ProfileFields = {
  delimiter: string
  has_header: boolean
  skip_rows: number
  date_column: number
  date_format: DateFormat
  amount_mode: 'single' | 'debit_credit'
  amount_column: number | null
  debit_column: number | null
  credit_column: number | null
  invert_sign: boolean
  description_column: number
  memo_column: number | null
}

export type ImportProfile = ProfileFields & {
  id: number
  name: string
  account_id: number | null
}

export type ProfileInput = ProfileFields & { name: string; account_id: number | null }

export type PreviewRow = { date: string; amount_cents: number; description: string; memo: string }
export type Problem = { line: number; reason: string }
export type Preview = { columns: string[]; rows: PreviewRow[]; problems: Problem[] }

export type Disposition = 'import' | 'skip' | 'match'

export type StagedRow = {
  id: number
  row_index: number
  date: string
  amount_cents: number
  raw_description: string
  raw_memo: string
  payee_id: number | null
  new_payee_name: string | null
  category_id: number | null
  memo: string | null
  disposition: Disposition
  is_duplicate: boolean
  applied_rule_id: number | null
  match: {
    transaction_id: number
    date: string
    amount_cents: number
    payee_id: number | null
    memo: string | null
  } | null
  bill: { occurrence_id: number; name: string; due_date: string } | null
  link_bill: boolean
  created_transaction_id: number | null
}

export type RowPatch = Partial<
  Pick<
    StagedRow,
    'payee_id' | 'new_payee_name' | 'category_id' | 'memo' | 'disposition' | 'link_bill'
  >
>

export type ImportBatch = {
  id: number
  account_id: number
  filename: string
  format: 'csv' | 'ofx' | 'qfx'
  profile_id: number | null
  status: 'staged' | 'committed' | 'undone'
  row_count: number
  imported_count: number
  duplicate_count: number
  matched_count: number
  parse_errors: string[]
  created_at: string
  committed_at: string | null
  undone_at: string | null
}

export type ImportBatchDetail = ImportBatch & { rows: StagedRow[] }
export type BatchResult = { batch: ImportBatch; balances: Balance[] }

export const DEFAULT_PROFILE: ProfileFields = {
  delimiter: ',',
  has_header: true,
  skip_rows: 0,
  date_column: 0,
  date_format: 'MM/DD/YYYY',
  amount_mode: 'single',
  amount_column: 2,
  debit_column: null,
  credit_column: null,
  invert_sign: false,
  description_column: 1,
  memo_column: null,
}

/** Largest file the server takes (D-078). */
export const MAX_FILE_BYTES = 5 * 1024 * 1024

export function isOfx(filename: string, text: string): boolean {
  return /\.(ofx|qfx)$/i.test(filename) || /<OFX>|OFXHEADER/i.test(text.slice(0, 4096))
}

export function fetchProfiles(signal?: AbortSignal): Promise<{ items: ImportProfile[] }> {
  return apiFetch('/import-profiles', { signal })
}

export function createProfile(body: ProfileInput): Promise<ImportProfile> {
  return apiFetch('/import-profiles', { method: 'POST', body })
}

export function updateProfile(id: number, body: ProfileInput): Promise<ImportProfile> {
  return apiFetch(`/import-profiles/${id}`, { method: 'PUT', body })
}

export function deleteProfile(id: number): Promise<void> {
  return apiFetch(`/import-profiles/${id}`, { method: 'DELETE' })
}

export function previewCsv(
  content: string,
  profile: ProfileFields,
  signal?: AbortSignal,
): Promise<Preview> {
  return apiFetch('/imports/preview', { method: 'POST', body: { content, profile }, signal })
}

export function stageImport(body: {
  account_id: number
  filename: string
  content: string
  profile_id: number | null
}): Promise<ImportBatchDetail> {
  return apiFetch('/imports', { method: 'POST', body })
}

export function fetchImports(
  accountId: number,
  signal?: AbortSignal,
): Promise<{ items: ImportBatch[] }> {
  return apiFetch(`/imports?account_id=${accountId}`, { signal })
}

export function fetchImport(id: number, signal?: AbortSignal): Promise<ImportBatchDetail> {
  return apiFetch(`/imports/${id}`, { signal })
}

export function patchRow(batchId: number, rowId: number, body: RowPatch): Promise<StagedRow> {
  return apiFetch(`/imports/${batchId}/rows/${rowId}`, { method: 'PATCH', body })
}

export function commitImport(id: number): Promise<BatchResult> {
  return apiFetch(`/imports/${id}/commit`, { method: 'POST' })
}

export function undoImport(id: number): Promise<BatchResult> {
  return apiFetch(`/imports/${id}/undo`, { method: 'POST' })
}

export function discardImport(id: number): Promise<void> {
  return apiFetch(`/imports/${id}`, { method: 'DELETE' })
}

export function applyRules(id: number): Promise<ImportBatchDetail> {
  return apiFetch(`/imports/${id}/apply-rules`, { method: 'POST' })
}
